### ISSUE TITLE
Idea: Map and filter at once

### ISSUE THREAD

**Issue by nickluger** [remeda/remeda#258]
Often, you do something like `array.map(...).filter(...)`, which iterates the whole array twice. To avoid that, you'd need to reach out to `reduce`.

It'd be great if there would be a function that does both in the same call.

Radash has such a thing: https://radash-docs.vercel.app/docs/getting-started#select

**Comment by TkDodo** [comment]
Are you looking for `flatMap` ?

**Comment by nickluger** [comment]
`flatMap` does map and flatten (i.e. map and then spread inner arrays into the outer array), AFAIK. 

I was thinking about map and **filter**, e.g.
 
```ts
// instead of this
people.filter(person => person.age >= 25).map(person => person.name)
// do this
people.filterMap(person => person.age >=25, person => person.name)
```

But it's not a necessity for me, as I don't have to deal with very big arrays currently. So there's no performance benefit. Might be different for others.

Was just an idea, i jotted down here, as I came across the Radash function.

**Comment by nickluger** [comment]
>  by returning an empty array when you want to remove something

Oh yes, a nice workaround 👍,  similar to `reduce`, but maybe a bit easier to read.
```js
people.reduce((acc, p) => p.age >= 25 ? [...acc, p.name] : acc, []);
```
I'll close this as the workaround seems succint and sufficient.

**Comment by Gregoor** [comment]
Heya, I use `.flatMap` for this use case as well but I think a `.filterMap()` would be more succinct. Then again, it would probably require introducing `None` symbol (to allow inclusion of null/undefined-values as a filter result) which might make this still out of scope?

**Comment by eranhirsch** [comment]
When using remeda's pipe operator, it runs the map and filter lazily so that the array isn't iterated over twice:

```ts
const mappedAndFilters = pipe(
  DATA,
  map(mapper),
  filter(predicate),
);
```

This will effectively do the same thing that a dedicated `maybeMap` function would do, with the added advantage that you can extend the processing pipeline further and filter by something other than `undefined` (which is what `maybeMap` would filter against).

I have a tentative plan to try and push some of the logic from Remeda's pipe directly into the utility functions so that we can run lazy evaluations even without the pipe and then doing:

```ts
const mappedAndFilters = filter(map(DATA, mapper), predicate);
```

will also be as efficient. Once that is done, there won't be a need for a standalone `maybeMap` anymore.
